knpu/episode3/Challenges/AbstractShip/ExtendAbstractClassCoding.php

```diff
@@ -0,0 +1,228 @@
+
+    public function getFileBuilder()
+    {
+        $fileBuilder = new FileBuilder();
+        $fileBuilder
+            ->addFileContents('AbstractDeathStar.php', <<<EOF
+EOF
+            )
+            ->addFileContents('DeathStar.php', <<<EOF
+<?php
+
+class DeathStar
+{
+    private \$crewSize;
+
+    private \$weaponPower;
+
+    public function setCrewSize(\$numberOfPeople)
+    {
+        \$this->crewSize = \$numberOfPeople;
+    }
+
+    public function getCrewSize()
+    {
+        return \$this->crewSize;
+    }
+
+    public function setWeaponPower(\$power)
+    {
+        \$this->weaponPower = \$power;
+    }
+
+    public function getWeaponPower()
+    {
+        return \$this->weaponPower;
+    }
+
+    public function makeFiringNoise()
+    {
+        echo 'BOOM x '.\$this->weaponPower;
+    }
+}
+EOF
+            )
+            ->addFileContents('DeathStarII.php', <<<EOF
+<?php
+
+class DeathStarII extends DeathStar
+{
+    // this DeathStar must *always* have 5000 people on it
+    public function getCrewSize()
+    {
+        return 5000;
+    }
+
+    public function makeFiringNoise()
+    {
+        echo 'SUPER BOOM';
+    }
+}
+EOF
+            )
+            ->addFileContents('index.php', <<<EOF
+<?php
+
+require 'AbstractDeathStar.php';
+require 'DeathStar.php';
+require 'DeathStarII.php';
+
+\$deathStar1 = new DeathStar();
+\$deathStar1->setCrewSize(3000);
+\$deathStar1->setWeaponPower(500);
+
+\$deathStar2 = new DeathStarII();
+\$deathStar2->setWeaponPower(999);
+
+?>
+
+<table>
+    <tr>
+        <td>&nbsp;</td>
+        <th>DeathStar 1</th>
+        <th>DeathStar 2</th>
+    </tr>
+    <tr>
+        <th>Crew Size</th>
+        <td><?php echo \$deathStar1->getCrewSize(); ?></td>
+        <td><?php echo \$deathStar2->getCrewSize(); ?></td>
+    </tr>
+    <tr>
+        <th>Weapon Power</th>
+        <td><?php echo \$deathStar1->getWeaponPower(); ?></td>
+        <td><?php echo \$deathStar2->getWeaponPower(); ?></td>
+    </tr>
+    <tr>
+        <th>Fire!</th>
+        <td><?php echo \$deathStar1->makeFiringNoise(); ?></td>
+        <td><?php echo \$deathStar2->makeFiringNoise(); ?></td>
+    </tr>
+</table>
+EOF
+            )
+            ->setEntryPointFilename('index.php')
+        ;
+
+        return $fileBuilder;
+    }
+
+    public function getExecutionMode()
+    {
+        return self::EXECUTION_MODE_PHP_NORMAL;
+    }
+
+    public function setupContext(CodingContext $context)
+    {
+    }
+
+    public function grade(CodingExecutionResult $result)
+    {
+        try {
+            $abstractClass = new \ReflectionClass('\AbstractDeathStar');
+        } catch (\Exception $e) {
+            throw new GradingException($e->getMessage().'. Did you create it?');
+        }
+        if (!$abstractClass->isAbstract()) {
+            throw new GradingException('The `AbstractDeathStar` class should be declared as abstract.');
+        }
+        if (false
+            or !$abstractClass->hasMethod('setCrewSize')
```

Comment:
this is perhaps too tricky, but `AbstractDeathstar` doesn't need `setCrewSize` - because `DeathstarII` just has a hardcoded crew size, so doesn't need this setter. Perhaps we should have a second check after this that checks to make sure that there is *not* a `getCrewSize` in `AbstractDeathstar`, and then clearly explain why it might not be needed.